applications/GeoMechanicsApplication/tests/cpp_tests/test_builder_and_solver_factory.cpp

```diff
@@ -0,0 +1,63 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Richard Faasse
+//
+
+#include "testing/testing.h"
+#include "custom_utilities/builder_and_solver_factory.hpp"
+#include "spaces/ublas_space.h"
+#include "linear_solvers/linear_solver.h"
+
+using namespace Kratos;
+using SparseSpaceType = UblasSpace<double, CompressedMatrix, Vector>;
+using LocalSpaceType = UblasSpace<double, Matrix, Vector>;
+using LinearSolverType = LinearSolver<SparseSpaceType, LocalSpaceType>;
+
+namespace Kratos::Testing
+{
+
+KRATOS_TEST_CASE_IN_SUITE(CreateBuilderAndSolver_ReturnsCorrectType_WhenBlockBuilderIsTrue, KratosGeoMechanicsFastSuite)
+{
+    const std::string validBlockParameters = R"(
+    {
+        "block_builder": true
+    }
+    )";
+
+    const auto solver = std::make_shared<LinearSolverType>();
+
+    const auto builder_and_solver = BuilderAndSolverFactory<SparseSpaceType, LocalSpaceType, LinearSolverType>::Create(Parameters{validBlockParameters}, solver);
+    const auto block_builder = dynamic_cast<const ResidualBasedBlockBuilderAndSolver<SparseSpaceType, LocalSpaceType, LinearSolverType>*>(builder_and_solver.get());
+    KRATOS_EXPECT_NE(block_builder, nullptr);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(CreateBuilderAndSolver_ReturnsCorrectType_WhenBlockBuilderIsFalse, KratosGeoMechanicsFastSuite)
+{
+    const std::string validBlockParameters = R"(
+    {
+        "block_builder": false
+    }
+    )";
+
+    const auto solver = std::make_shared<LinearSolverType>();
+    const auto builder_and_solver = BuilderAndSolverFactory<SparseSpaceType, LocalSpaceType, LinearSolverType>::Create(Parameters{validBlockParameters}, solver);
+    const auto elimination_builder = dynamic_cast<const ResidualBasedEliminationBuilderAndSolver<SparseSpaceType, LocalSpaceType, LinearSolverType>*>(builder_and_solver.get());
+    KRATOS_EXPECT_NE(elimination_builder, nullptr);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(CreateBuilderAndSolver_Throws_WhenBlockBuilderIsUndefined, KratosGeoMechanicsFastSuite)
+{
+    const auto solver = std::make_shared<LinearSolverType>();
+    const auto factory = BuilderAndSolverFactory<SparseSpaceType, LocalSpaceType, LinearSolverType>();
+    KRATOS_EXPECT_EXCEPTION_IS_THROWN(factory.Create(Parameters{}, solver),
+                                      "the block_builder parameter is not defined, aborting BuilderAndSolverCreation")
```

Comment:
Since `Create` is a static member function, you'd better write it like this (just as for the two above test cases):
```suggestion
    KRATOS_EXPECT_EXCEPTION_IS_THROWN(BuilderAndSolverFactory<SparseSpaceType, LocalSpaceType, LinearSolverType>::Create(Parameters{}, solver),
                                      "the block_builder parameter is not defined, aborting BuilderAndSolverCreation")
```

Reply:
Yes, you're right. However, you may overcome the above problem by wrapping the first argument of the macro in a pair of parenthesis, i.e.:
```suggestion
    const auto factory = BuilderAndSolverFactory<SparseSpaceType, LocalSpaceType, LinearSolverType>();
    KRATOS_EXPECT_EXCEPTION_IS_THROWN((BuilderAndSolverFactory<SparseSpaceType, LocalSpaceType, LinearSolverType>::Create(Parameters{}, solver)),
                                      "the block_builder parameter is not defined, aborting BuilderAndSolverCreation")
```